XIPs/xip-31-message-history.md

```diff
@@ -0,0 +1,435 @@
+
+1. Load the Backup Account File from disk.
+1. Create a temporary client instance with a new location for the database.
+1. Call `receive` on the temporary client instance to download all messages stored on the network.
+1. Convert all messages into a Message Backup File.
+1. Import the Message Backup File into the main account's database.
+1. Delete the database associated with the temporary client instance and the Message Backup File.
+
+While some apps may choose to support Backup Account Files directly, we can also offer means to convert Backup Account Files into Remote Message Backups. This allows developers to support all possible backup scenarios simply by supporting the more common Remote Message Backups.
+
+#### Converting a Backup Account File into a Remote Message Backup
+
+XMTP Labs should create a simple web app to convert Backup Account Files into Remote Message Backups. In this app, a user could import their Backup Account File (with the file never leaving their machine) and create a temporary client instance with that backup file. The user would be presented with any outstanding `MessageHistoryBackupRequests` and select any backup requests they want to fulfill. The client would download all unread messages from the network and proceed with the regular Message Backup Provider flow for Remote Message Backups. All data would be cleared from the client as soon as the operation was completed.
+
+This same web app could be used to create Backup Account Files for cases where a user's preferred XMTP app does not support creating Backup Account Files. This means every user on the XMTP network can create a Backup Account File if desired.
+
+### Backup Storage Provider
+
+A Backup Storage Provider is a simple HTTP service with two endpoints. Anyone can implement a Backup Storage Provider. It is up to the Backup Requester app to choose the Backup Service Provider for their app.
+
+These are the required APIs for a minimal Backup Storage Provider:
+
+`POST /backups`:
+Request body would contain the file as multipart/form-data.
+
+Example response:
+
+```json
+{
+  "downloadUrl": "https://backupproviderdomain.com/backups/some-long-unguessable-download-id"
+}
+```
+
+`GET /backups/$DOWNLOAD_ID`:
+Returns the uploaded file matching the ID
+
+It would be the responsibility of the Backup Storage Provider to authenticate requests to `/backups` and mitigate abuse. Uploaded files would only need to be stored for maybe 72 hours before they could be safely purged, as backups are meant to be temporary storage. We could also just delete the file after it had been downloaded once.
+
+XMTP Labs would provide a reference implementation of a Backup Storage Provider.
+
+I am also proposing that XMTP Labs run a Backup Storage Provider as a free public good for the next two years, at which point this functionality would become part of an ecosystem of third-party gateway service providers.
```

Comment:
OK to leave this in?

Reply:
yes, we are already running this service